### ISSUE TITLE
@sc-core tag exceeds maximum length of 255 chars

### ISSUE THREAD

**Issue by jtriley** [jtriley/StarCluster#348]
From the mailing list:

http://star.mit.edu/cluster/mlarchives/1974.html

I have had limited success getting Starcluster to successfully launch a
cluster with EC2-VPC nodes under the development version (0.9999). Using a
certain AMI I can easily launch a Starcluster cluster with EC2-VPC nodes,
but using a different AMI it fails to launch.  I do set the config
variables "VPC_ID" and "SUBNET_ID" and the only difference between the two
cluster templates is the AMI that is used.

Both AMIs used successfully launch a Starcluster cluster with EC2-classic
nodes.  The only noted difference between the AMIs is that the one that
successfully launches a Starcluster cluster with VPC-EC2 nodes is a
private AMI that is "shared" with the account that I am running my VPC
within.  The AMI that doesn't work with Starcluster-VPC is one that is
private AMI "owned" by the account I am running my VPC within.   

I believe the error I am getting has something to do with the Tags,
specifically the "@sc-core" tag's value being beyond 255 characters, but I
could be wrong.  Below I have included an example of the successful
launch, the failed launch (including error message), and the listed
clusters after both commands.

Any suggestions on how to address this issue would be greatly appreciated.

Thanks in advance for the help,

-Jennifer

Below is what it looks like when I have a successful launch

```
(starcluster)root@xxxxxxxxxxx:~# starcluster start -c testvpcA vpcA
StarCluster - ([1]http://star.mit.edu/cluster) (v. 0.9999)
Software Tools for Academics and Researchers (STAR)
Please submit bug reports to [2]starcluster@mit.edu
>>> Validating cluster template settings...
>>> Cluster template settings are valid
>>> Starting cluster...
>>> Launching a 1-node cluster...
>>> Creating security group @sc-vpcA...
Reservation:r-2843fa4e
>>> Waiting for cluster to come up... (updating every 30s)
>>> Waiting for all nodes to be in a 'running' state...
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Waiting for SSH to come up on all nodes...
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Waiting for cluster to come up took 1.574 mins
>>> The master node is
>>> Configuring cluster...
>>> Running plugin starcluster.clustersetup.DefaultClusterSetup
>>> Configuring hostnames...
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Creating cluster user: sgeadmin (uid: 1007, gid: 1000)
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Configuring scratch space for user(s): sgeadmin
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Configuring /etc/hosts on each node
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Starting NFS server on master
>>> Setting up NFS took 0.113 mins
>>> Configuring passwordless ssh for root
>>> Configuring passwordless ssh for sgeadmin
>>> Running plugin starcluster.plugins.sge.SGEPlugin
>>> Configuring SGE...
>>> Setting up NFS took 0.000 mins
>>> Removing previous SGE installation...
>>> Installing Sun Grid Engine...
>>> Creating SGE parallel environment 'orte'
1/1 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
100%
>>> Adding parallel environment 'orte' to queue 'all.q'
>>> Configuring cluster took 0.679 mins
>>> Starting cluster took 2.307 mins
The cluster is now ready to use. To login to the master node
as root, run:
    $ starcluster sshmaster vpcA
If you're having issues with the cluster you can reboot the
instances and completely reconfigure the cluster from
scratch using:
    $ starcluster restart vpcA
When you're finished using the cluster and wish to terminate
it and stop paying for service:
    $ starcluster terminate vpcA
Alternatively, if the cluster uses EBS instances, you can
use the 'stop' command to shutdown all nodes and put them
into a 'stopped' state preserving the EBS volumes backing
the nodes:
    $ starcluster stop vpcA
WARNING: Any data stored in ephemeral storage (usually /mnt)
will be lost!
You can activate a 'stopped' cluster by passing the -x
option to the 'start' command:
    $ starcluster start -x vpcA
This will start all 'stopped' nodes and reconfigure the
cluster.
```

Below is what it looks like when I have a FAILED launch

```
(starcluster)root@xxxxxxxxxxx:~# starcluster start -c testvpcB vpcB
StarCluster - ([3]http://star.mit.edu/cluster) (v. 0.9999)
Software Tools for Academics and Researchers (STAR)
Please submit bug reports to [4]starcluster@mit.edu
>>> Validating cluster template settings...
>>> Cluster template settings are valid
>>> Starting cluster...
>>> Launching a 1-node cluster...
>>> Creating security group @sc-vpcB...
!!! ERROR - InvalidParameterValue: Tag value exceeds the maximum length of
255 characters
Traceback (most recent call last):
  File "/root/.virtualenvs/starcluster/starcluster/starcluster/cli.py",
line 274, in main
    sc.execute(args)
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/commands/start.py",
line 220, in execute
    validate_running=validate_running)
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
1537, in start
    return self._start(create=create, create_only=create_only)
  File "<string>", line 2, in _start
  File "/root/.virtualenvs/starcluster/starcluster/starcluster/utils.py",
line 111, in wrap_f
    res = func(*arg, **kargs)
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
1552, in _start
    self.create_cluster()
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
1066, in create_cluster
    self._create_flat_rate_cluster()
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
1091, in _create_flat_rate_cluster
    force_flat=True)[0]
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
859, in create_nodes
    cluster_sg = [5]self.cluster_group.name
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
657, in cluster_group
    self._add_tags_to_sg(sg)
  File
"/root/.virtualenvs/starcluster/starcluster/starcluster/cluster.py", line
698, in _add_tags_to_sg
    sg.add_tag(static.CORE_TAG, core_settings)
  File
"/root/.virtualenvs/starcluster/local/lib/python2.7/site-packages/boto-2.19.0-py2.7.egg/boto/ec2/ec2object.py",
line 82, in add_tag
    dry_run=dry_run
  File
"/root/.virtualenvs/starcluster/local/lib/python2.7/site-packages/boto-2.19.0-py2.7.egg/boto/ec2/connection.py",
line 4026, in create_tags
    return self.get_status('CreateTags', params, verb='POST')
  File
"/root/.virtualenvs/starcluster/local/lib/python2.7/site-packages/boto-2.19.0-py2.7.egg/boto/connection.py",
line 1158, in get_status
    raise self.ResponseError(response.status, response.reason, body)
EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidParameterValue</Code><Message>Tag
value exceeds the maximum length of 255
characters</Message></Error></Errors><RequestID>1f589605-8f30-472d-8989-22ea120aea14</RequestID></Response>
```

When if FAILS it creates only a security group see "listclusters" below

```
(starcluster)root@xxxxxxxxxxx:~# starcluster listclusters
StarCluster - ([6]http://star.mit.edu/cluster) (v. 0.9999)
Software Tools for Academics and Researchers (STAR)
Please submit bug reports to [7]starcluster@mit.edu
-------------------------------
vpcB (security group: @sc-vpcB)
-------------------------------
Launch time: N/A
Uptime: N/A
Zone: N/A
Keypair: N/A
EBS volumes: N/A
Cluster nodes: N/A
-------------------------------
vpcA (security group: @sc-vpcA)
-------------------------------
Launch time: 2013-12-10 14:39:36
Uptime: 0 days, 00:04:23
Zone: us-east-1b
Keypair: Starcluster_VPC
EBS volumes: N/A
Cluster nodes:
     master running i-1d745b65 10.0.0.138
Total nodes: 1
```

**Comment by ghost** [comment]
I think I figured out the problem and I have a solution that works for creating a starcluster under the VPC BUT you can't change its attributes after creation other than doing a force termination.

Starcluster works in EC2-Classic because they don't limit the Tag Values to a certain size, but with EC2-VPC Tag Values are limited to size 255 characters. I think this was discussed before if you use the following link https://github.com/jtriley/StarCluster/issues/21  and search for words “3) StarCluster + VPC limitation:"  you can read more about this issue.  

Basically the @sc-core Tag's value can be greater than 255 characters in length, if this is the case then Starcluster software will fail to create a cluster under the VPC.  This is why for my one AMI (had @sc-core Tag value less than 255 characters) I could successfully create a starcluster under the VPC, but for all other AMI's I wasn't able to.

Temporary Solution:
1 ) Locate the .../starcluster/cluster.py file and open it up in your favorite editor
2) Go to line 681 this should be the first line of the method "_add_tags_to_sg"  and it should have following text "def _add_tags_to_sg(self, sg):"
3) The value for 'static.CORE_TAG' is the Tag Name '@sc-core' and it's value is 'core_settings'.  The issue occurs in adding this Tag with line 698 'sg.add_tag(static.CORE_TAG, core_settings)'. If the value of 'core_settings' is greater than 255 then this is where the program fails.  Add the following 3 lines of code (indicated by (+)) between lines 697 'if not static.CORE_TAG in sg.tags:' and 698 'sg.add_tag(static.CORE_TAG, core_settings)'
-------------------------------- Code Change -------------------------------------------------------------------------------
697           if not static.CORE_TAG in sg.tags:
698(+)            if(len(core_settings) > 255):
699(+)               print "\nWarning: For ", static.CORE_TAG, " truncating core_settings from ", len(core_settings), " to length 255."
700(+)               core_settings=core_settings[:255]
701                sg.add_tag(static.CORE_TAG, core_settings)
-------------------------------- Code Change -------------------------------------------------------------------------------

WARNING - there are implications of making this code change.  In doing so you will not be able to change the cluster other than a forced termination IF the core_settings value was truncated.

Seems like the @sc-core Tag holds cluster settings for a cluster that have been serialized and then compressed and the tag value is that compressed value.  Truncating this value allows you to create a cluster BUT you can't change attributes once it is created because the software needs a non-truncated value to link back to the cluster (I think?).

Once up and running the cluster runs fine, but ultimately for Starcluster to work consistently with EC2-VPC the code would need to be changed as to limit the @sc-core value to only 255 characters or fewer.  Maybe eliminating some of the options in the core tag, better compression algorithm, or ?? might be an easy fix to keep from having to do a big re-write of the code.  My temporary solution works for a 'static' cluster under the VPC.

Thanks,
-Jennifer

**Comment by jtriley** [comment]
@jmstaab Yep, that's the issue. The reason I asked for the config is I'm curious what is making the value that long. I have a more general fix for this but in the mean time would you mind pasting your `[cluster]` section? Thanks!

**Comment by ghost** [comment]
I have a feeling you were about to suggest the fix that I have outlined below.  It involves no code change just a change in how you set up your cluster template such that your @sc-core tag squeezes within the 255 character limit.

Things to note about EC2-VPC and the this issue: 
1)  EC2-VPC limits their tag lengths to 255 characters (this is not true for EC2-Classic)
2)  @sc-core tag and (non-truncated) value are needed for starcluster to function properly
3)  @sc-core tag values are serialized (via json) and compressed to make them usable but reduce their size
4)  changing zlib.compress() function to use level 9 (highest compression) instead of default 6 doesn't solve the problem

SOLUTION
Changing your [cluster]/cluster template configuration in the configuration file such that 'master_instance_type' and 'node_instance_type' are of same type (like m1.small) AND 'master_image_id' and 'node_image_id' use the same ID (like ami-xxxxxxxx) will allow compression such that the @sc-core tag value can be compressed to a size that is under 255 characters.  

You can use VPC even if your master and worker nodes are of different types or use different AMI's. To allow this type of variation and not violate the 255 character limit, I would suggest creating a cluster template with the AMI id and type you want for the master node for both master and (worker) node.  Then use starcluster addnode  with -i and -I options to add all the worker nodes you want of the TYPE and AMI ID you want.  I have tried it and it seems to work perfectly.

Just in case someone runs into a similar problem in the future, below are the dictionary of values that are serialized and compressed to create the @sc-core tag's value:

{'disable_cloudinit': False, 'master_instance_type': 'm1.small', 'disable_queue': False, 'subnet_id': 'subnet-xxxxxxxx', 'availability_zone': 'us-east-1b', 'node_image_id': 'ami-xxxxxxxx', 'node_instance_type': 'm1.small', 'use_nfs_crossmnt': True, 'cluster_size': 1, 'master_image_id': 'ami-xxxxxxxx', 'dns_prefix': False}

Thanks,
-Jennifer

**Comment by ghost** [comment]
I have a feeling you were about to suggest the fix that I have outlined
below. It involves no code change just a change in how you set up your
cluster template such that your @sc-core tag squeezes within the 255
character limit.

Things to note about EC2-VPC and the this issue:
1) EC2-VPC limits their tag lengths to 255 characters (this is not true for
EC2-Classic)
2) @sc-core tag and (non-truncated) value are needed for starcluster to
function properly
3) @sc-core tag values are serialized (via json) and compressed to make
them usable but reduce their size
4) changing zlib.compress() function to use level 9 (highest compression)
instead of default 6 doesn't solve the problem

SOLUTION
Changing your [cluster]/cluster template configuration in the configuration
file such that 'master_instance_type' and 'node_instance_type' are of same
type (like m1.small) AND 'master_image_id' and 'node_image_id' use the same
ID (like ami-xxxxxxxx) will allow compression such that the @sc-core tag
value can be compressed to a size that is under 255 characters.

You can use VPC even if your master and worker nodes are of different types
or use different AMI's. To allow this type of variation and not violate the
255 character limit, I would suggest creating a cluster template with the
AMI id and type you want for the master node for both master and (worker)
node. Then use starcluster addnode with -i and -I options to add all the
worker nodes you want of the TYPE and AMI ID you want. I have tried it and
it seems to work perfectly.

Just in case someone runs into a similar problem in the future, below are
the dictionary of values that are serialized and compressed to create the
@sc-core tag's value:

{'disable_cloudinit': False, 'master_instance_type': 'm1.small',
'disable_queue': False, 'subnet_id': 'subnet-xxxxxxxx',
'availability_zone': 'us-east-1b', 'node_image_id': 'ami-xxxxxxxx',
'node_instance_type': 'm1.small', 'use_nfs_crossmnt': True, 'cluster_size':
1, 'master_image_id': 'ami-xxxxxxxx', 'dns_prefix': False}

Thanks,
-Jennifer

**Comment by jtriley** [comment]
@jmstaab No I wasn't about to suggest that actually - I was talking about modifying the code to fix this in general so that users arent trying to fit the tags - users ideally should not even be aware of this unless they're contributing to starcluster development.

Again, if you could _please_ paste your config that would be useful. Also, are you using the official StarCluster or some other fork? I'm seeing `use_nfs_crossmnt` in your tag's value which does not exist in the official code. This is likely why you're running over the limit.

**Comment by jtriley** [comment]
@jmstaab OK just googling I found `use_nfs_crossmnt` in @rqbanerjee's fork of StarCluster. This option isn't officially supported and is almost certainly the reason you're hitting this error. In any case I'm going to try a more robust approach to storing core settings in tags that uses multiple tags to increase the limit.

**Comment by ghost** [comment]
Yes we are using his fork, I forgot that the "USE_NFS_CROSSMNT" wasn't standard.  Regardless, things seem to be working with some minor adjustments to how one uses the cluster templates.  If you still want config let me know.  It will take a bit to remove all the sensitive information.

Thanks,
-Jennifer

**Comment by rqbanerjee** [comment]
Sorry, I was unaware of the VPC 255 character tag limit. Let me know if there's anything I can change about that commit. Guess one cannot always assume that there is extra space available. Sorry to give you a headache, Jennifer.

**Comment by ghost** [comment]
Don't worry about it, I have a method that allows it all to work without changing code and I know more about the inner workings of Starcluster than I did a few days ago. 

Sorry to have caused a flurry of activity/undo worry about an unsupported change in a fork of Starcluster. 

-Jennifer

**Comment by rqbanerjee** [comment]
If you remove the line about use_nfs_crossmnt from the config, it'll go away from the tags when the cluster starts up.

**Comment by nitecoder** [comment]
Any update on the fix? We just ran into this issue while working with 0.95 vanilla release.
I would be happy to post the config, except that we've broken it up into a number of files. Do I need to zip them all up or is there a way to get starcluster to split out unified config as it sees it?

**Comment by ghost** [comment]
I haven't used the new 0.95 release yet, so I can't comment on the issue.  One simple thing you may try that might help get under the 255 character compression limit without changing any of the Starcluster code is the following.

In your Starcluster configuration file use a cluster template, where your 'cluster_size' = 1 cluster and the  'master_instance_type' and 'node_instance_type' are of same type (like m1.xlarge) AND 'master_image_id' and 'node_image_id' use the same ID (like ami-xxxxxxxx).  Once this cluster is up just start the worker nodes with 'addnode' using 'node_instance_type' and 'node_image_id' that you want (assuming you want them different than the master node).  Setting all these features the same in the cluster template might allow for more compression of the @sc-core tag value so that it might not exceed 255 characters limit.

Good Luck!

-Jennifer

**Comment by jtriley** [comment]
@nitecoder Please try the following and paste the output here (be careful with any sensitive info):

```
$ pip install ipython
$ starcluster shell
[1]> cfg.clusters['smallcluster']
(press enter key)
```

Replace `smallcluster` with the name of your cluster template that you're using. That should give us more info where it's spilling over.

**Comment by nitecoder** [comment]
@jtriley  Thanks! Surprisingly I didn't even need a running cluster to get this. Here's the info:

I am including two configs. First one is from a cluster that is still running. It was created with a snapshot of development branch of Starcluster from a few months ago. The second one is the one that won't start, using 0.95 version. Otherwise, I think the config is more or less the same between them. 

Looks like 0.95 is including a lot more of the config into the tag. Is that possible?

Version that works - snapshot from development branch a few months ago:

```
{'__name__': 'cluster spot-prod-us-west-1b',
 'availability_zone': 'us-west-1b',
 'cluster_shell': 'bash',
 'cluster_size': 2,
 'cluster_user': 'sgeadmin',
 'disable_cloudinit': False,
 'disable_queue': False,
 'dns_prefix': False,
 'extends': 'spot-abstract',
 'force_spot_master': False,
 'key_location': 'keys/starcluster-us-west-1.rsa',
 'keyname': 'starcluster-us-west-1',
 'master_image_id': 'ami-02674b47',
 'master_instance_type': 'm1.medium',
 'node_image_id': 'ami-02674b47',
 'node_instance_type': 'c1.xlarge',
 'node_instance_types': [],
 'permissions': [],
 'plugins': [],
 'public_ips': True,
 'spot_bid': 0.5,
 'subnet_id': 'subnet-XXXXXXXX',
 'userdata_scripts': [],
 'volumes': []}
```

Current version 0.95 - runs in to 255 characters limit.

```
{'__name__': 'cluster autotagger-prod-us-west-1b_2',
 'availability_zone': 'us-west-1b',
 'cluster_shell': 'bash',
 'cluster_size': 1,
 'cluster_user': 'sgeadmin',
 'disable_cloudinit': False,
 'disable_queue': False,
 'dns_prefix': False,
 'extends': 'spot-prod-us-west-1b',
 'force_spot_master': False,
 'key_location': 'keys/starcluster-us-west-1.rsa',
 'keyname': 'starcluster-us-west-1',
 'master_image_id': 'ami-02674b47',
 'master_instance_type': 'm3.large',
 'node_image_id': 'ami-02674b47',
 'node_instance_type': 'm2.4xlarge',
 'node_instance_types': [],
 'permissions': {'flask': {'__name__': 'flask',
   'cidr_ip': '0.0.0.0/0',
   'from_port': 8000,
   'ip_protocol': 'tcp',
   'to_port': 8000},
  'http': {'__name__': 'http',
   'cidr_ip': '0.0.0.0/0',
   'from_port': 80,
   'ip_protocol': 'tcp',
   'to_port': 80}},
 'plugins': [{'__name__': 'plugin autotagger-debs-remove',
   'packages': 'libopenblas-base, libopenblas-dev',
   'setup_class': 'package_remover.PackageRemover'},
  {'__name__': 'plugin autotagger-debs',
   'packages': 'g++, python-dev, libatlas-dev, liblapack-dev, gfortran, swig, libatlas-base-dev',
   'setup_class': 'starcluster.plugins.pkginstaller.PackageInstaller'}],
 'public_ips': True,
 'spot_bid': 0.5,
 'subnet_id': 'subnet-XXXXXXXX',
 'userdata_scripts': [],
 'volumes': {'v-tagger-uswest1b2': {'device': None,
   'mount_path': '/data',
   'partition': None,
   'volume_id': 'vol-YYYYYYYY'}}}
```

**Comment by jtriley** [comment]
@nitecoder Perfect, thanks! I'm looking into this now.

**Comment by nitecoder** [comment]
Hi Justin, any update on this? 
It seems that we should be splitting this among multiple tags, kind of like the old trick of storing info in multiple browser cookies. Is that the approach you are taking?

**Comment by CarlosBorroto** [comment]
Got hit by this. It goes away if I choose not to have different instance types for master and nodes.

My config:

`{'__name__': 'cluster gpscience',
 'availability_zone': 'us-east-1d',
 'cluster_shell': 'bash',
 'cluster_size': 1,
 'cluster_user': 'sgeadmin',
 'disable_cloudinit': False,
 'disable_queue': True,
 'dns_prefix': True,
 'extends': None,
 'force_spot_master': False,
 'key_location': 'keys/id_rsa.gpscience',
 'keyname': 'gpscience',
 'master_image_id': 'ami-3354545a',
 'master_instance_type': 'm1.small',
 'node_image_id': 'ami-6b211202',
 'node_instance_type': 'cr1.8xlarge',
 'node_instance_types': [],
 'permissions': [],
 'plugins': [{'__name__': 'plugin sge',
   'master_is_exec_host': 'False',
   'setup_class': 'starcluster.plugins.sge.SGEPlugin'}],
 'public_ips': True,
 'spot_bid': 0.5,
 'subnet_id': None,
 'userdata_scripts': [],
 'volumes': {'gpscience': {'device': None,
   'mount_path': '/mnt',
   'partition': None,
   'volume_id': 'vol-746eba36'}}}`

**Comment by jtriley** [comment]
Fixed in 0.95.1: http://star.mit.edu/cluster/docs/latest/changelog.html#version-0-95-1
